src/jitlayers.cpp

```diff
@@ -1629,6 +1629,12 @@ JuliaOJIT::JuliaOJIT()
     DLSymOpt(std::make_unique<DLSymOptimizer>(false)),
     ContextPool([](){
         auto ctx = std::make_unique<LLVMContext>();
+        if (!ctx->hasSetOpaquePointersValue())
+#ifndef JL_LLVM_OPAQUE_POINTERS
+            ctx->setOpaquePointers(false);
+#else
+            ctx->setOpaquePointers(true);
+#endif
```

Comment:
Can you wrap this into a function and export it:
```
JL_DLLEXPORT_CODEGEN extern "C"
LLVMContextRef *JuliaLLVMContextCreate(void)
{
...
}
```
Or perhaps a more complete option would be to export this ability as:
```
JL_DLLEXPORT_CODEGEN extern "C"
LLVMModuleRef *JuliaLLVMModuleCreateWithName(const char *ModuleID)
{
     // return a new Module in a new LLVMContext compatible with the current JIT target
     // (it may be illegal to use the result for AOT targets or external targets like GPUs)
     // the caller is responsible for adding locks as needed and disposing of the Module and LLVMContext when finished
    ...
}
```
since then we can set it up with all the required properties (TargetMachine, etc)

Reply:
That's a good idea, but it doesn't cover the GPU use case, where we may want to use a different opaque pointer setting. That's why I was thinking about making it a task-local setting (which apparently doesn't work because the task that compiles isn't guaranteed to be the same as the one that invokes the compiler).

Basically, we need to be able to:

```julia
@generated function emit_something()
   # creates LLVM.Context, serializes to IR to llvmcall.
   # opaque pointer setting of this context should respect the compiler context
end

# when calling from Julia, use the same opaque pointer setting as the Julia JIT
emit_something()

# when calling from e.g. GPUCompiler with a different opaque-pointer setting
# it should respect that
ctx = LLVM.Context(; opaque_pointers=...)
mod = LLVM.Module(ctx)
ccall(:jl_create_native, ..., (emit_something, mod))
```

Reply:
For more challenging cases (like llvmcalls that could not auto-convert to untyped upon parsing), we can keep that in a separate module and just be sure to pass that to the JIT first. They can't be compiled together, but they can be linked together

Reply:
> They can't be compiled together, but they can be linked together

GPUCompiler links at the LLVM level, i.e., within a single context (because most GPU back-ends don't have a real JIT). You can't deserialize untyped IR in an typed context, or vice-versa (it asserts in BitcodeReader), so I don't think this would work.

Reply:
LLVM claimed that you could deserialize typed IR in any context (https://llvm.org/docs/OpaquePointers.html#opaque-pointers-mode). Was that not true?

Reply:
From that link, AFAICT, all we really need to do here is force the global default to be typed-pointers, and then override that per-context where we know it is valid for the backend, since converting typed->opaque is legal (and relatively trivial) while going the other direction is trickier.

Reply:
> LLVM claimed that you could deserialize typed IR in any context (https://llvm.org/docs/OpaquePointers.html#opaque-pointers-mode).

Ah, I did not know that, and had only run into the reverse not working. It seems to work as expected:

```
julia> typed = """
       define void @typed(i64 zeroext %0) {
       top:
         %memcpy_refined_dst = inttoptr i64 %0 to i64*
         store i64 1, i64* %memcpy_refined_dst, align 1
         ret void
       }
       """

julia> ctx = Context(; opaque_pointers=true)
LLVM.Context(0x000056401d55be30, opaque ptrs)

julia> mod = parse(LLVM.Module, typed)
# works!
```

```
julia> opaque = """
       define void @opaque(i64 zeroext %0) {
       top:
         %coercion = inttoptr i64 %0 to ptr
         store i64 1, ptr %coercion, align 1
         ret void
       }
       """

julia> ctx = Context(; opaque_pointers=false)
LLVM.Context(0x000056401d6526e0, typed ptrs)

julia> mod2 = parse(LLVM.Module, opaque)
warning: ptr type is only supported in -opaque-pointers mode
  %coercion = inttoptr i64 %0 to ptr
                                 ^
ERROR: LLVM error: error: expected type
  %coercion = inttoptr i64 %0 to ptr
                                 ^
```

So defaulting to typed pointers in LLVM.jl should be fine, I think. I'll experiment some more tomorrow.